sphinx-mode.el

```diff
@@ -36,26 +40,61 @@
   '((t (:inherit fixed-pitch)))
   "Face used for code blocks.")
 
-(defun sphinx-fontify-code-block (limit)
-  "Fontify code blocks from point to LIMIT."
-  (condition-case nil
-      (while (re-search-forward "\.\. code-block:: \\(.*?\\)\n\n\\( +\\)" limit t)
-        (let* ((block-start (match-end 0))
-               (block-highlight-start (match-beginning 2))
-               (lang (match-string 1))
-               (prefix (match-string 2))
-               (prefix-search (concat "^" prefix))
-               block-end)
-          (while (progn
-                   (forward-line)
-                   (looking-at prefix-search)))
-          (setq block-end (point))
-          (sphinx-src-font-lock-fontify-block lang block-start block-end)
-          (add-face-text-property
-           block-highlight-start block-end
-           'sphinx-code-block-face 'append)))
-    (error nil)))
+(defvar sphinx-code-block-default-language "python"
+  "The default language used for highlighting code blocks.")
 
+(defun sphinx--fontify-code-blocks (limit)
+  "Fontify code blocks from point to LIMIT."
+  (save-excursion
+    ;; This is for the ‘.. code-block::’ syntax.
+    (while (re-search-forward "\.\. code-block::" limit t)
```

Comment:
I think this was also mistake in my code but as I understand font-lock now we should only work on one match at a time and not loop here.  font-lock will then loop the matcher until it runs after the limit.

I wrote a blog post about it https://fuco1.github.io/2017-06-11-Font-locking-with-custom-matchers.html but this code is older than that.